README.md
# NAME

ZMQx::Class - DEPRECATED - OO Interface to ZMQ

# VERSION

version 0.007

# SYNOPSIS

DEPRECATED - This was only a prototype and never used in production. I doubt it still works with current zmq.

    # a ZeroMQ publisher
    # see example/publisher.pl
    use ZMQx::Class;
    my $publisher = ZMQx::Class->socket( 'PUB', bind => 'tcp://*:10000' );

    while ( 1 ) {
        my $random = int( rand ( 10_000 ) );
        say "sending $random hello";
        $publisher->send( [ $random, 'hello' ] );
        select( undef, undef, undef, 0.1);
    }


    # a ZeroMQ subscriber
    # see example/subscriber.pl
    use ZMQx::Class;
    use Anyevent;

    my $subscriber = ZMQx::Class->socket( 'SUB', connect => 'tcp://localhost:10000' );
    $subscriber->subscribe( '1' );

    my $watcher = $subscriber->anyevent_watcher( sub {
        while ( my $msg = $subscriber->receive ) {
            say "got $msg->[0] saying $msg->[1]";
        }
    });
    AnyEvent->condvar->recv;

# DESCRIPTION

DEPRECATED - This was only a prototype and never used in production. I doubt it still works with current zmq. But here are the old docs:

`ZMQx::Class` provides an object oriented & Perlish interface to [ZeroMQ](http://www.zeromq.org/) 3.2. It builds on [ZMQ::FFI](https://metacpan.org/module/ZMQ::FFI).

Before you use `ZMQx::Class`, please read the excellent <ZeroMQ Guide|http://zguide.zeromq.org>. It's a fun and interesting read, containing everything you need to get started with ZeroMQ, including lots of example code.

# METHODS

## context

    my $ctx = ZMQx::Class->context;

Return the current context for this process.

ZMQx::Class pools one context per process (pid). It will always use the currents process socket, and set up a new context after a fork. This should be very helpful for most use cases (e.g. when you need a 0mq socket inside a preforked webapp).

If you want to get a new context, use `ZMQx::Class->_new_context`. But you will have to manage this context yourself!

## socket

    my $socket = ZMQx::Class->socket( $type );
    my $socket = ZMQx::Class->socket( $type, bind => $endpoint );
    my $socket = ZMQx::Class->socket( $type, connect => $endpoint, \%opts );
    my $socket = ZMQx::Class->socket( $context, $type );

`socket` is a factory method that returns a new `ZMQx::Class::Socket` object. The new socket will use the default process-wide context.

    my $socket = ZMQx::Class->socket( $type );

Returns a new socket of the given type. Types are valid 0mq sockets:

    REQ REP DEALER ROUTER PULL PUSH PUB SUB XPUB XSUB PAIR

This socket will be neither bound nor connected, and have no sockopts set.

**connect / bind**

    my $socket = ZMQx::Class->socket( $type, [bind | connect] => $endpoint );

If you call socket like this, you will get back a socket that's already bound or connected to the given endpoint.

**options**

    my $socket = ZMQx::Class->socket( $type, connect => $endpoint, {
        sndhwm  => 500
    });

You can pass in a hashref containing valid sockopts to set them before connect/bind. A lot of sockopts can only be set before connect/bind, so if you want to set them, either pass them to `socket`, or do not use the autoconnect/bind feater.

See Sockopts in ZMQx::Class::Socket for a list of valid options.

**context**

If you really need to, you can pass a context object as the first agrument to `socket`.

    my $socket = ZMQx::Class->socket( $context, $type );

# SEE ALSO

[ZMQ](https://metacpan.org/module/ZMQ::Socket) is another perlish interface to libzmq. We found it to still be a bit too low-level (eg, you still need to import the sockopt-constants). ZMQ support libzmq 2 and 3, and promises a single interface to both. We are only interested in libzmq 3. ZMQx::Class includes some smartness to handle setup of context objects, even across forks. Lastly, we found this note from the docs not very promising: "Personally, I'd recommend only using this module for your one-shot scripts, and use ZMQ::LibZMQ\* for all other uses."

# THANKS

Thanks to [Validad](http://www.validad.com/) for sponsoring the development of this module.

# AUTHOR

Thomas Klausner <domm@plix.at>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 - 2015 by Validad AG.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for ZMQx::Class

{{$NEXT}}

0.008	2021-01-10T19:07:32
    - DEPRECATED!


0.007 Wed Oct 29 20:37:30 PDT 2014
    - String type message decode / encode utf8

0.006   2014-10-02T03:57:49+0200
    - added task distribution example (David Schmidt)
    - Only close open sockets in close(), updated dependency reference from ZMQ::LibZMQ3 to ZMQ::FFI (Michael Kröll)
    - lazy_pirate socket reset (Thomas Klausner)
    - Confess instead of die, cleanup (Klaus Ita)
    - Better diagnostics when (failing to) decode payloads garbled on the wire. (Nicholas Clark)
    - Flag-day re-ordering of on_error's parameters to be consistent with return's. (Nicholas Clark)
    - Change rpc_bind() from an exported subroutine to a method. (Nicholas Clark)
    - A much simpler structure for creating client/server tests. (Nicholas Clark)
    - Refactor the RPC test to avoid using timers. (Nicholas Clark)
    - Fix the RPC::Message::Response escape hatch. (Nicholas Clark)
    - Handle ROUTER sockets properly. (Nicholas Clark)
    - coverted Loop.pm to parametric role; store allowed comands in dispatch table; passthrough mode via payload=>raw (Thomas Klausner)
    - finished message/header objects, loop, some tests (Thomas Klausner)
    - added ZMQx::RPC::Message with first test (Thomas Klausner)
    - Ensure that send() always passes strings as sequences of UTF-8 to ZMQ::FFI (Nicholas Clark)
    - Add a send_bytes method, for when you really need to send raw bytes. (Nicholas Clark)
    - Update send() to reflect the (non-)return value from ZMQ::FFI::send(). (Nicholas Clark)
    - Create a proper test case showing the UTF-8 / zmq issue (Klaus Ita)
    - Call the correct ->socket  (RT63106_Building_deb_for_0mq_4) (Klaus Ita)
    - ZMQx::Class for ZMQ::FFI module (Klaus Ita)
    - added ZMQ_LAST_ENDPOINT helper (Thomas Klausner)


0.005   2013-08-05T15:57:14+0200
    - removed ZMQx::Class::AnyEvent in favour of $socket->anyevent_watcher
      (Thomas Klausner)
    - implemented $socket->send($string) (Thomas Klausner)
    - croak if cannot bind/connect (Thomas Klausner)
    - inproc push/pull (Thomas Klausner)
    - better fork safety (Klaus Ita)
    - use fd internally (Klaus Ita)
    - started with docs, added a pub/sub example (Thomas Klausner)
    - deprecated receive_multipart(), replaced by receive() (Thomas Klausner)
    - removed  wait_for_message (suggested by Marc Lehmann) (Thomas Klausner)
    - removed receive_all_multipart_messages (Thomas Klausner)
    - proper & working implementation of receive_multipart (Thomas Klausner)
    - aded get_events (Thomas Klausner)


0.004   2013-02-19T22:07:43+0100
    - warn when trying to set a sockopt after connecting (Thomas Klausner)
    - pass sockopts to ZMQx::Class->socket (Thomas Klausner)
    - ZMQx::Class->socket now works without $context (gets it from singleton)


0.003   2013-01-18T09:43:30+0100
    - smarter send method that can handle single, multipart and
      flags (Thomas Klausner)
    - added tests (Thomas Klausner)
    - added get_sockopt helper methods (Thomas Klausner)
    - use zmq_msg_send instead of zmq_sendmsg (Thomas Klausner)
    - new_context, context (Klaus Ita)
    - added wait_for_message (Thomas Klausner)


0.002   2013-01-07T15:13:24+0100
    - create set_SOCKOPT shortcut methods
    - use random port numbers to avoid clashes in tests
    - added ZMQx::Class->context


0.001
    - started new dist



lib/ZMQx/Class.pm
package ZMQx::Class;

# ABSTRACT: DEPRECATED - OO Interface to ZMQ
# VERSION

warn __PACKAGE__ .' is DEPRECATED, please do not use this module anymore';

use strict;
use warnings;
use 5.010;
use ZMQx::Class::Socket;
use Carp qw(croak carp);

my $__CONTEXT = {};

use ZMQ::FFI;
use ZMQ::Constants qw(
    ZMQ_DEALER
    ZMQ_PAIR
    ZMQ_PUB
    ZMQ_PULL
    ZMQ_PUSH
    ZMQ_REP
    ZMQ_REQ
    ZMQ_ROUTER
    ZMQ_SUB
    ZMQ_XPUB
    ZMQ_XSUB
);

my %types = (
    'REQ'    => ZMQ_REQ,
    'REP'    => ZMQ_REP,
    'DEALER' => ZMQ_DEALER,
    'ROUTER' => ZMQ_ROUTER,
    'PULL'   => ZMQ_PULL,
    'PUSH'   => ZMQ_PUSH,
    'PUB'    => ZMQ_PUB,
    'SUB'    => ZMQ_SUB,
    'XPUB'   => ZMQ_XPUB,
    'XSUB'   => ZMQ_XSUB,
    'PAIR'   => ZMQ_PAIR,
);

sub _new_context {
    my $class = shift;
    return ZMQ::FFI->new( @_ );
}

=method context

    my $ctx = ZMQx::Class->context;

Return the current context for this process.

ZMQx::Class pools one context per process (pid). It will always use the currents process socket, and set up a new context after a fork. This should be very helpful for most use cases (e.g. when you need a 0mq socket inside a preforked webapp).

If you want to get a new context, use C<< ZMQx::Class->_new_context >>. But you will have to manage this context yourself!

=cut

sub context {
    my $class = shift;
    return $__CONTEXT->{$$} //= $class->_new_context(@_);
}

=method socket

    my $socket = ZMQx::Class->socket( $type );
    my $socket = ZMQx::Class->socket( $type, bind => $endpoint );
    my $socket = ZMQx::Class->socket( $type, connect => $endpoint, \%opts );
    my $socket = ZMQx::Class->socket( $context, $type );

C<socket> is a factory method that returns a new C<ZMQx::Class::Socket> object. The new socket will use the default process-wide context.

    my $socket = ZMQx::Class->socket( $type );

Returns a new socket of the given type. Types are valid 0mq sockets:

    REQ REP DEALER ROUTER PULL PUSH PUB SUB XPUB XSUB PAIR

This socket will be neither bound nor connected, and have no sockopts set.

B<connect / bind>

    my $socket = ZMQx::Class->socket( $type, [bind | connect] => $endpoint );

If you call socket like this, you will get back a socket that's already bound or connected to the given endpoint.

B<options>

    my $socket = ZMQx::Class->socket( $type, connect => $endpoint, {
        sndhwm  => 500
    });

You can pass in a hashref containing valid sockopts to set them before connect/bind. A lot of sockopts can only be set before connect/bind, so if you want to set them, either pass them to C<socket>, or do not use the autoconnect/bind feater.

See Sockopts in ZMQx::Class::Socket for a list of valid options.

B<context>

If you really need to, you can pass a context object as the first agrument to C<socket>.

    my $socket = ZMQx::Class->socket( $context, $type );

=cut

sub socket {
    my $class           = shift;
    my $context_or_type = shift;
    my ( $context, $type );
    if ( ref($context_or_type) =~ /^ZMQ::FFI::ZMQ\d::Context/ ) {
        $context = $context_or_type;
        $type    = shift;
    }
    else {
        $context = $class->context;
        $type    = $context_or_type;
    }
    my ( $connect, $address, $opts ) = @_;
    croak "no such socket type: $type" unless defined $types{$type};

    my $socket = ZMQx::Class::Socket->new(
        _socket => $context->socket( $types{$type} ),
        type    => $type,
        _pid    => $$,
        _init_opts_for_cloning => [ $class, $type, @_ ],
    );

    if ($opts) {
        while ( my ( $opt, $val ) = each %$opts ) {
            my $method = 'set_' . $opt;
            if ( $socket->can($method) ) {
                $socket->$method($val);
            }
            else {
                carp "no such sockopt $opt";
            }
        }
    }

    if ( $connect && $address ) {
        if ( $connect eq 'bind' ) {
            $socket->bind($address);
        }
        elsif ( $connect eq 'connect' ) {
            $socket->connect($address);
        }
        else {
            croak "no such connect type: $connect";
        }
    }
    return $socket;
}

q{ listening to: Tosca - Odeon };

__END__

=head1 SYNOPSIS

DEPRECATED - This was only a prototype and never used in production. I doubt it still works with current zmq.

  # a ZeroMQ publisher
  # see example/publisher.pl
  use ZMQx::Class;
  my $publisher = ZMQx::Class->socket( 'PUB', bind => 'tcp://*:10000' );

  while ( 1 ) {
      my $random = int( rand ( 10_000 ) );
      say "sending $random hello";
      $publisher->send( [ $random, 'hello' ] );
      select( undef, undef, undef, 0.1);
  }


  # a ZeroMQ subscriber
  # see example/subscriber.pl
  use ZMQx::Class;
  use Anyevent;

  my $subscriber = ZMQx::Class->socket( 'SUB', connect => 'tcp://localhost:10000' );
  $subscriber->subscribe( '1' );

  my $watcher = $subscriber->anyevent_watcher( sub {
      while ( my $msg = $subscriber->receive ) {
          say "got $msg->[0] saying $msg->[1]";
      }
  });
  AnyEvent->condvar->recv;

=head1 DESCRIPTION

DEPRECATED - This was only a prototype and never used in production. I doubt it still works with current zmq. But here are the old docs:

C<ZMQx::Class> provides an object oriented & Perlish interface to L<ZeroMQ|http://www.zeromq.org/> 3.2. It builds on L<ZMQ::FFI|https://metacpan.org/module/ZMQ::FFI>.

Before you use C<ZMQx::Class>, please read the excellent <ZeroMQ Guide|http://zguide.zeromq.org>. It's a fun and interesting read, containing everything you need to get started with ZeroMQ, including lots of example code.

=head1 SEE ALSO

L<ZMQ|https://metacpan.org/module/ZMQ::Socket> is another perlish interface to libzmq. We found it to still be a bit too low-level (eg, you still need to import the sockopt-constants). ZMQ support libzmq 2 and 3, and promises a single interface to both. We are only interested in libzmq 3. ZMQx::Class includes some smartness to handle setup of context objects, even across forks. Lastly, we found this note from the docs not very promising: "Personally, I'd recommend only using this module for your one-shot scripts, and use ZMQ::LibZMQ* for all other uses."

=head1 THANKS

Thanks to L<Validad|http://www.validad.com/> for sponsoring the development of this module.

